src/SA/ldapsearch/entry.c

```diff
@@ -0,0 +1,398 @@
+    Time.QuadPart = Time.QuadPart - TICKSTO1970;
+    Time.QuadPart = (uint64_t)Time.QuadPart / (uint64_t)TICKSPERSEC;
+    dwSeconds = Time.u.LowPart;
+
+    return dwSeconds;
+}
+#endif
+VOID ReturnDateTime(DWORD dwSeconds)
+{
+    LARGE_INTEGER Time;
+    FILETIME FileTime;
+    SYSTEMTIME SystemTime;
+    PCHAR DateBuffer[80];
+    PCHAR TimeBuffer[80];
+
+    Time.QuadPart = ((LONGLONG)dwSeconds * TICKSPERSEC) + TICKSTO1970;
+    
+    FileTime.dwLowDateTime = Time.u.LowPart;
+    FileTime.dwHighDateTime = Time.u.HighPart;
+    KERNEL32$FileTimeToLocalFileTime(&FileTime, &FileTime);
+    KERNEL32$FileTimeToSystemTime(&FileTime, &SystemTime);
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   DATE_SHORTDATE,
+                   &SystemTime,
+                   NULL,
+                   DateBuffer,
+                   ARRAYSIZE(DateBuffer));
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   TIME_NOSECONDS,
+                   &SystemTime,
+                   NULL,
+                   TimeBuffer,
+                   ARRAYSIZE(TimeBuffer));
+
+    internal_printf("%S %S", DateBuffer, TimeBuffer);
+}
+
+PCHAR GetDomainControllerHostName(){
+	DWORD dwRet;
+	PDOMAIN_CONTROLLER_INFO pdcInfo;
+
+	dwRet = NETAPI32$DsGetDcNameA(NULL, NULL, NULL, NULL, 0, &pdcInfo);
+	if (ERROR_SUCCESS == dwRet) {
+		internal_printf("[*] Using DC: %s\n", pdcInfo->DomainControllerName);		
+	} else {
+ 	    BeaconPrintf(CALLBACK_ERROR, "Failed to identify PDC, are we domain joined?");
+	}
+
+	PCHAR hostName = pdcInfo->DomainControllerName + 2; // Remove preceeding backslashes
+
+	NETAPI32$NetApiBufferFree(pdcInfo);
+	return hostName;
+}
+
+LDAP* InitialiseLDAPConnection(PCHAR hostName, PCHAR distinguishedName){
+	LDAP* pLdapConnection = NULL;
+
+    pLdapConnection = WLDAP32$ldap_init(hostName, 389);
+    
+    if (pLdapConnection == NULL)
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Failed to establish LDAP connection on 389.");
+        WLDAP32$ldap_unbind(pLdapConnection);
+    }
+    else {
+		internal_printf("[+] Initialised Connection.\n");
+    }
+
+	//////////////////////////////
+	// Bind to DC
+	//////////////////////////////
+    ULONG lRtn = 0;
+
+    lRtn = WLDAP32$ldap_bind_s(
+                pLdapConnection,      // Session Handle
+                distinguishedName,                // Domain DN
+                NULL,                 // Credential structure
+                LDAP_AUTH_NEGOTIATE); // Auth mode
+
+    if(lRtn == LDAP_SUCCESS)
+    {
+    	internal_printf("[+] Bind Successful.\n");
+    }
+    else
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Bind Failed: %i", lRtn);
+        WLDAP32$ldap_unbind(pLdapConnection);
+    }
+    return pLdapConnection;
+}
+
+LDAPMessage* ExecuteLDAPQuery(LDAP* pLdapConnection, PCHAR distinguishedName, char * ldap_filter, char * ldap_attributes){
+	ULONG errorCode = LDAP_SUCCESS;
+    LDAPMessage* pSearchResult;
+    
+	if(strlen(ldap_attributes)!=0){
```

Comment:
For this there are a few options.

First ldap_attributes should be passed in as NULL if none are given.  This will remove having to do the strlen check.  You can see how that is handled in some of the other BOFs in this repo such as netuser, line 215.

second, this currently limits us to either one attribute or all attributes.  It would be good if the user could pass a comma separated list of attributes, and then you allocate the char ** here and populate it.  If you decide to keep it to one attribute your PCHAR [] should be of size 2.  with the second element being NULL.  That is what would make it always be a NULL terminated array.

Reply:
First - updated, made cna more concise as a result 👍 

Second - updated to take multiple attributes using strtok() function. Seems to work on 64bit, but i'm running into an issue where this function produces a crash in 32-bit beacon. I've added a directive as a stop gap. Have you experienced anything similar in other BOFs? Any ideas what the cause could be?